Add `concatenate_datasets` for iterable datasets

lhoestq opened the pull request:
`concatenate_datasets` currently only supports lists of `datasets.Dataset`, not lists of `datasets.IterableDataset` like `interleave_datasets`

Fix https://github.com/huggingface/datasets/issues/2564

I also moved `_interleave_map_style_datasets` from combine.py to arrow_dataset.py, since the logic depends a lot on the `Dataset` object internals

And I moved `concatenate_datasets` from arrow_dataset.py to combine.py to have it with `interleave_datasets` (though it's also copied in arrow_dataset module for backward compatibility for now)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_4500). All of your documentation changes will be reflected on that endpoint.

mariosasko reviewed (commented):
Looks good already!

There is a slight difference in `concatenate_datasets` between the version for map-style datasets and the one for iterable datasets:
* if `axis=0`, the map-style version checks the feature types (`features` can be `None` in iterable datasets, so it's ok not to have this check) of the shared columns, but **doesn't require the equal set of column names among the input datasets**, i.e., the following works:
  ```python
   >>> from datasets import *
   >>> a = Dataset.from_dict({"a": [1, 2]})
   >>> b = Dataset.from_dict({"b": ["aa", "bb"]})
   >>> concatenate_datasets([a, b])
   ```
   We need to address this.
* if `axis=1`, the map-style version checks the length of input datasets  (besides the column names check for duplicates) and throws an error if the lengths are not equal. Somewhat expected that this is ignored for iterable datasets, so it can stay as-is IMO.

Two nits in code:

lhoestq commented:
Thanks ! I addressed your comments :)

> There is a slight difference in concatenate_datasets between the version for map-style datasets and the one for iterable datasets

Indeed, here is what I did to fix this:

- axis 0: fill missing columns with None.
(I first iterate over the input datasets to infer their columns from the first examples, then I set the features of the resulting dataset to be the merged features)

- axis 1: **fill the missing rows with None**, for consistency with axis 0
(but let me know what you think, I can still revert this behavior and raise an error when one of the dataset runs out of examples)
We might have to align the non-streaming concatenation with this behavior though, for consistency. What do you think ?

mariosasko commented on src/datasets/iterable_dataset.py:
```diff
@@ -587,6 +673,11 @@ def _iter_shard(self, shard_idx: int):
 
     def _apply_feature_types(self, example):
         if self.features:
+            example = dict(example)
+            # add missing columns
+            for column_name in self.features:
+                if column_name not in example:
+                    example[column_name] = None
```
If I'm not mistaken, one can accidentally introduce new columns in `cast` and these columns' values will be returned as `None` due to this change. But I think it's OK to leave this as-is for now and add checks later for consistency with `Dataset`.

mariosasko reviewed (commented):
Cool!

> We might have to align the non-streaming concatenation with this behavior though, for consistency. What do you think ?

Yes, we can address that in a subsequent PR

lhoestq commented on src/datasets/iterable_dataset.py:
```diff
@@ -170,6 +172,85 @@ def shard_data_sources(self, shard_idx: int) -> "CyclingMultiSourcesExamplesIter
+                except StopIteration:
+                    ex_iterators.remove(ex_iterator)
+            if ex_iterators:
+                if i == 0:
+                    _check_column_names([column_name for example in examples for column_name in example])
```
I'd keep it because I think `HorizontallyConcatenatedMultiSourcesExamplesIterable` is also responsible for checking that there are no duplicates. This is also tested at

https://github.com/huggingface/datasets/blob/94b229377df59a0ff6d9fb93cad3c9644fb0a871/tests/test_iterable_dataset.py#L506-L508

mariosasko replied:
That's also OK :).

lhoestq commented:
Added more comments as suggested, and some typing

While factorizing _apply_features_types for both IterableDataset and TypedExamplesIterable, I  fixed a missing `token_per_repo_id` that was not passed to TypedExamplesIteable

mariosasko reviewed (approved):
Looks good! Thanks!
